### ISSUE TITLE
simplify sort

### ISSUE THREAD

**Issue by alexbyk** [json-api/json-api#324]
How about to simplify sort to `?sort=name:-1`

full example:
`?sort=name:1,price:-1` which means (order does matter) sort by:
1. by `name` asc
2. by `price` desc

Benefits:
1) In this form it should be simple to split all pairs by `,` character, then split pairs to key:value
2) Right now it's impossible to use `-field` in the DB. But the are no words about field names should not begin from the `-` character. So current implementation of sorting makes api inconsistent 

We can also make it backward compatible. But it would be simpler to drop current syntax out as inconsistent

**Comment by wvteijlingen** [comment]
I don't see any benefits to this form over the current spec. The current spec uses a comma as a separator as well, so splitting is equally easy. I don't think an attribute name starting with a dash would be common. This form actually has the same problems, what happens when I have a field named `name:1`?

Perhaps some constraints on the attribute name would resolve this problem, otherwise I would leave sorting as is.

**Comment by alexbyk** [comment]
if you have a field named `name:1`, you could sort by `name:1:-1`|`name:1:1` without problems, and you gave an example why not to use a short form `?sort=name` => `?sort=name:1`. But you also gave a perfect example why current impl. should be changed (except a short form from my proposal). If you have 2 fields `-name` and `name` it would be a real problem in the current impl.

About benefits. `<>` with (1,0,-1) operations looks simpler than getting  charAt(0) and checking if it equals to `-` in the most of languages

**Comment by wvteijlingen** [comment]
If I have `?sort=name:-1`, and two fields `name` and `name:-1`, we cannot differentiate between an descending sort on the field `name` and a ascending sort on the field`name:-1`. It is the same problem as with `name` and `-name`. The only upside is that a field named `name:-1` would be even more uncommon.

> About benefits. <> with (1,0,-1) operations looks simpler than getting charAt(0) and checking if it equals to - in the most of languages.

I don't think we should base our decisions on whether something looks simpler in a specific language. Of course we should make parsing as easy as possible, but I don't think this would be a real improvement.

Maybe others disagree with me on this though :).

**Comment by alexbyk** [comment]
@wvteijlingen i wrote that `?sort=name` without `:1|-1` is optional. I edited a comment (cleared optional confusing section). If you have 2 fields, the sorting query would be `?sort=name:-1:1,name:-1` which means sort by `name:-1` asc then by `name` desc. And the only remaining problem is `,` in fields names

**Comment by wvteijlingen** [comment]
Ok, If you remove the optionality (and force clients to state asc or desc) then you won't have this problem indeed. Still, I don't think this is something we need to worry about for 1.0. :)

**Comment by alexbyk** [comment]
Or, in other hand, we should pay more attention and time before releasing 1.0 and call it api because it doesn't pass this simple test (it's already a second critical problem i've found, the first is https://github.com/json-api/json-api/issues/305) from the real world. But the good part of github is everyone can have his own opinion

Offtopik:
You don't have to force clients to state asc or desc. My first suggestion with `sort='name'` worked well all times, I removed the optionality only to make an example a little simpler (because if it's not obvious for someone, it could be not obvious for others) and to keep discussion in right direction (away from asking me how to differ options containing ':')

**Comment by ahacking** [comment]
For inspiration one could have a look at the oasis standards/specifications for [odata](http://www.odata.org/documentation/odata-version-4-0/) as its quite a solid spec. The [$orderBy](http://docs.oasis-open.org/odata/odata/v4.0/errata01/os/complete/part1-protocol/odata-v4.0-errata01-os-part1-protocol-complete.html#_Toc399426801) query option covers ordering.

**Comment by alexbyk** [comment]
Right now I prefer to consider this not only an arch. bug, but also a security issue in the most of the cases. Especially with no-schema DBs.

Also you can see I didn't suggest something new, I suggested to keep things simple, because most DBs already use a sort syntax the same way. 

@ahacking thanks for the link

**Comment by lgebhardt** [comment]
I would rather we tighten up the spec by putting some constraints on valid characters for field names, rather than change the sort format. It seems that is really the underlying issue.

I know JSON-API will not only be backed by a SQL database, but it's probably the most common case. In SQL starting a field name with a '-' is not valid, unless it's quoted. I do not favor supporting quoting of field names in this spec. If we can leave quoting out I think we end up with [0-9,a-z,A-Z$_](basic Latin letters, digits 0-9, dollar, underscore). That would SQL compliant, and probably compliant with non-sql databases as well. Then the current sort format would be satisfactory. In addition it keeps the implementation simple and avoids messy escaping rules.

**Comment by alexbyk** [comment]
@lgebhardt what exactly SQL do u mean? In mysql `-field` is a valid fieldname, how to use it in sql query - is the other story (the same way as other fields, `select `foo`, `-bar` from table`)

Sometimes it is good idea to limit available names, but in this case it'll look very funny, especially if someone will try to answer "why fields have such weird constraints"

**Comment by alexbyk** [comment]
I decided to provide some code snippets to show, why this is 'keeping api simple'

In backend, 2 functions, commented `ugly` covers all cases including when fields are named 'ugly:1', `simple` doesn't respect that but is simpler and can be copypasted to any language

``` perl
use Modern::Perl;
use experimental 'signatures';
use Data::Dumper;

#sub ugly($str) {
#  [map { /(.+):(-?1)$/ && ($1, $2); } split ',', $str];
#}
#
## 'foo:1' => -1, bar => 1
#my $arr2 = ugly("foo:1:-1,bar:1");
#say Dumper $arr2;

sub simple($str) {
  [map { split ':' } split ',', $str];
}

# foo => -1, bar => 1
my $arr1 = simple("foo:-1,bar:1");
say Dumper $arr1;

```

In client (javascript page, how to render current page (for example showing sort by price or not))

``` javascript
var sort = {
  foo: -1
};
if (sort.foo === -1) {
  console.log('desc');
}
```

Much more clear than dealing with `-foo,bar,-baz`, can be written by everyone in no time

**Comment by lgebhardt** [comment]
@alexbyk, for MySQL `-field` is a valid name, but you must quote it in a SQL statement (http://dev.mysql.com/doc/refman/5.0/en/identifiers.html). So I was proposing we restrict field names to the characters that can be used in SQL in unquoted field names.

We could allow quoting of field names in the spec. This would allow the use of other characters. For example you could sort like this: `sort=-"-name"`. 

However many client and server side tools will also have limitations on field names since they will back the fields with methods in their respective languages. In my server side implementation of the spec, https://github.com/cerebris/jsonapi-resources, I don't think I'll be allowing the quoted version of names since it won't work without major changes to the code  (without an aliasing system I'm also limited to ruby restrictions on method names). In fact the '$' will not be allowed either. In JavaScript '-' isn't a valid character for a variable name. I'd rather avoid these problems with reasonable restrictions, or at least recommendations, in the spec.

Maybe difficulty on the implementation side of things shouldn't drive the spec, but it certainly seems reasonable to consider it.

**Comment by alexbyk** [comment]
@lgebhardt in JS `-` is a valid character for object's key. I'm against using ugly names starting with `-,_,$`, but I don't see a reason why to restrict this in API. See an example. MongoDB always has a `_id` field. So if tomorrow some new super cool db will decide to always have a '-foo' key, jsonApi will have a problems from nowhere. 

Quoiting  `sort=-"-name"` solves a problem, but I think it would be much more simpler to avoid that problem and have nothing to solve.

I see only one vote against it: if someone already use existing syntax and it'll be a problem to rewrite existing code

**Comment by alexbyk** [comment]
I'm done. Let's wait for other opinions)

**Comment by lgebhardt** [comment]
@alexbyk, in JS a variable name can't contain `-` (http://stackoverflow.com/questions/1661197/valid-characters-for-javascript-variable-names), but a JSON key can.

It won't be hard to rewrite the rules to parse the sort parameters, but I still don't see the point. However I'm not arguing against your proposed format, but rather that it brings up the need for some clarification on valid field names. Even if sort is changed I still think we should restrict fields from containing a '-'.

**Comment by alexbyk** [comment]
@lgebhardt i know that js variables can's start with `-`, but keys (and JSON keys can). I just don't see how this is related to our discussion at all (and how it could be a reason to restrict smthng):)

**Comment by kevin-brown** [comment]
I'm [-0](https://docs.djangoproject.com/en/dev/internals/contributing/bugs-and-features/#how-we-make-decisions) on this, mostly for the reasons specified in https://github.com/json-api/json-api/issues/324#issuecomment-65226890 and https://github.com/json-api/json-api/issues/324#issuecomment-65233432.

I personally don't see the point in making this change, but it's not something that I'd really fight back against if there is strong support.

I may not fully be remembering this part of the spec, but couldn't you just use `-name` and `--name` (similar to how in the proposed change you can use `name:1` and `name:1:-1`)?

**Comment by alexbyk** [comment]
Ok, Just for persons who will join this discussion later (and don't want to read many latters above)

Current api makes it impossible to work with db containing fields starting with `-` because of sort syntax . And can make a surprise for someone because it doesn't restrict that directly. I proposed a way to solve this, but looks like other vis-a-vis are against this proposal because it doesn't really matter

**Comment by lgebhardt** [comment]
@alexbyk, I brought up JS simply because it's likely that clients will implement models with attributes named the same as the fields. I'm not sure that should drive the rules, but will certainly drive the practical usage.

I think it's reasonable that the spec take an opinion on naming of fields, even if to say "anything goes as long as it's valid JSON". If '-' is allowed we will need a way of escaping or quoting it in the current sort rules. If your new proposal is adopted and ':' is allowed in a field name we will likewise need a way to escape or quote it. The spec precludes the dash character that would also solve the issue.

So to summarize we need to either prohibit the '-' or require escaping/quoting of the fields in a request. I would rather prohibit the '-', but I'm fine with either option.

**Comment by ahacking** [comment]
Both json-ld and odata use json keys starting with '@' and keys can also contain ':' as those specs support namespaced keys and schemas.

Json-api making restrictions on keys universally is only going to sideline the spec to a curiosity as it will never be able to carry arbitrary json from other systems.

Restricting keys would need to apply to the attribute and link keys in a resource only and not things nested within an attribute. But it seems kind of retarded to impose such a restriction because of the sorting convention chosen.

**Comment by ahacking** [comment]
The easiest fix for the spec would to simply require each sort key to always be prefixed with a '+' or a '-' for ascending and descending order respectively. This way the spec is agnostic to what the key starts with.

There remains escaping of other significant characters like commas or quotes within quoted strings.

**Comment by alexbyk** [comment]
@ahacking, If i'm correct, you proposed always have `?sort=+name1,-name2`. This makes sense with one exception: `+` has a special meaning and means whitespace in URL and should be encoded to `%2b`, which adds a little complexity, but at least saves a project from future heisenbugs too.

Right now my vote is +0 for my own proposal or for the @ahacking's idea.  (I decided to not use jsonApi in my project, because found it brings more problems for me, than solves, but I will watch it and wait, that's why not +1 but +0). They both solve a problem, but my proposal looks a little simpler to implement for me and can be used with no code and explanation for everyone (see my examples above)

**Comment by alexbyk** [comment]
I'm closing this issue (a reason is in prev paragraph, I lost an interest in it) right now, but I still think it's a real problem. So everyone who shares the same thoughts will be able to reopen it

Thanks for discussion

**Comment by ahacking** [comment]
@alexbyk 

I was assuming that all url components would be urlEncoded as a separate concern.

You are not alone with the watch and wait, and problems with the current spec.  Given what odata.org  and JSON-LD / Hydra provide already it is quite clear that those efforts are far further along than json-api.
